7.tcga-classify/classify.py

```diff
@@ -0,0 +1,233 @@
+"""
+2018 Gregory Way
+7.tcga-classify/classify.py
+
+Predict Ras and TP53 status in TCGA based on compressed expression features
+
+Usage:
+
+    python classify.py
+
+Output:
+Three DataFrames storing ROC, precision-recall, and classifier coefficients
+for every compression model trained in their ability to predict Ras pathway
+activation and TP53 pathway inactivation
+"""
+
+import os
+import glob
+import numpy as np
+import pandas as pd
+
+from scripts.tcga_util import (
+    get_threshold_metrics,
+    summarize_results,
+    extract_coefficients,
+    align_matrices,
+    process_y_matrix,
+    train_model,
+)
+
+np.random.seed(123)
+
+# Load constants
+genes = ["TP53", "RAS"]
+filter_prop = 0.05
+filter_count = 15
+folds = 5
+algorithms = ["pca", "ica", "nmf", "dae", "vae"]
+signals = ["signal", "shuffled"]
+alphas = [0.1, 0.13, 0.15, 0.2, 0.25, 0.3]
+l1_ratios = [0.15, 0.16, 0.2, 0.25, 0.3, 0.4]
+
+# Load data to build y matrices
+base_url = "https://github.com/greenelab/pancancer/raw"
+commit = "2a0683b68017fb226f4053e63415e4356191734f"
+
+# Load data
+file = "{}/{}/data/sample_freeze.tsv".format(base_url, commit)
+sample_freeze_df = pd.read_table(file, index_col=0)
+
+file = "{}/{}/data/pancan_mutation_freeze.tsv.gz".format(base_url, commit)
+mutation_df = pd.read_table(file, index_col=0)
+
+file = "{}/{}/data/copy_number_loss_status.tsv.gz".format(base_url, commit)
+copy_loss_df = pd.read_table(file, index_col=0)
+
+file = "{}/{}/data/copy_number_gain_status.tsv.gz".format(base_url, commit)
+copy_gain_df = pd.read_table(file, index_col=0)
+
+file = "{}/{}/data/mutation_burden_freeze.tsv".format(base_url, commit)
+mut_burden_df = pd.read_table(file, index_col=0)
+
+# What results to track
+full_metrics_list = []
+full_auc_list = []
+full_aupr_list = []
+full_coef_list = []
+all_failed_params = []
+
+# Obtain a list of locations for each feature matrix (X)
+z_matrix_dict = {}
+for signal in signals:
+    z_matrix_dict[signal] = {}
+
+    if signal == "signal":
+        results_dir = "TCGA_results"
+    else:
+        results_dir = "TCGA_shuffled_results"
+
+    matrix_dir = os.path.join(
+        "..", "2.ensemble-z-analysis", "results", results_dir, "ensemble_z_matrices"
+    )
+
+    for comp_dir in os.listdir(matrix_dir):
+        matrix_comp_dir = os.path.join(matrix_dir, comp_dir)
+        z_dim = comp_dir.split("_")[2]
+        z_matrix_dict[signal][z_dim] = {}
+
+        for z_file in glob.glob("{}/*_z_*".format(matrix_comp_dir)):
+            seed = os.path.basename(z_file).split("_")[1]
+
+            if seed not in z_matrix_dict[signal][z_dim].keys():
+                z_matrix_dict[signal][z_dim][seed] = {}
+
+            if "_test_" in z_file:
+                z_matrix_dict[signal][z_dim][seed]["test"] = z_file
```

Comment:
I haven't seen a data structure like this..is this a hypercube?  Or just a nested dictionary?

Reply:
nested dictionary. It is how I'm storing specific parameter combinations for the given z matrix